### PR TITLE
🏗️ Refactorings for `extension` controller

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area open-source
/kind cleanup

**What this PR does / why we need it**:
This PR refactors the Operator's `extension` controller, as a suggestion to improve maintenance, tests and readability. This is a preparation for further enhancements, e.g. applying extensions in the runtime cluster.

**Which issue(s) this PR fixes**:
Part of #9635

**Special notes for your reviewer**:
/cc @MartinWeindel @oliver-goetz

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/operator/controller/extension/reconciler.go [hunk 1]
```diff
@@ -156,20 +161,14 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, virtualClus
...
 
-	log.Info("Reconciling extension virtual resources")
-	if err := r.reconcileExtensionVirtualClusterResources(reconcileCtx, log, virtualClusterClientSet.Client(), extension); err != nil {
-		conditions.installed = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.installed, gardencorev1beta1.ConditionFalse, ConditionReconcileFailed, err.Error())
-		return errors.Join(err, r.updateExtensionStatus(ctx, log, extension, conditions))
-	}
-
-	log.Info("Reconciling admission virtual resources")
-	if err := r.reconcileAdmissionVirtualClusterResources(reconcileCtx, log, virtualClusterClientSet, extension); err != nil {
+	log.Info("Reconciling ControllerRegistration and ControllerDeployment")
```
Comment (MartinWeindel) [inline pkg/operator/controller/extension/reconciler.go]: Isn't "Reconciling" misleading here?
```suggestion
	log.Info("Deploying ControllerRegistration and ControllerDeployment")
```

File: pkg/operator/controller/extension/reconciler.go [hunk 2]
```diff
@@ -156,20 +161,14 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, virtualClus
...
+	if err := r.controllerRegistration.Reconcile(reconcileCtx, log, virtualClusterClientSet.Client(), extension); err != nil {
 		conditions.installed = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.installed, gardencorev1beta1.ConditionFalse, ConditionReconcileFailed, err.Error())
 		return errors.Join(err, r.updateExtensionStatus(ctx, log, extension, conditions))
 	}
 
-	log.Info("Reconciling admission runtime resources")
-	if err := r.reconcileAdmissionRuntimeClusterResources(reconcileCtx, log, genericTokenKubeconfigSecretName, extension); err != nil {
+	log.Info("Reconciling admission deployment")
```
Comment (MartinWeindel) [inline pkg/operator/controller/extension/reconciler.go]:
```suggestion
	log.Info("Deploying admission controller")
```

File: pkg/operator/controller/extension/admission/admission_test.go [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+		extension     *operatorv1alpha1.Extension
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		log = logger.MustNewZapLogger(logger.DebugLevel, logger.FormatJSON, logzap.WriteTo(GinkgoWriter))
+		ctrl = gomock.NewController(GinkgoT())
+
+		ociRefRuntime = "local-extension-runtime:v1.2.3"
+		ociRefVirtual = "local-extension-virtual:v1.2.3"
```
Comment (MartinWeindel) [inline pkg/operator/controller/extension/admission/admission_test.go]:
```suggestion
		ociRefRuntime = "admission-test-runtime:v1.2.3"
		ociRefApplication = "admission-test-application:v1.2.3"
```

File: pkg/operator/controller/extension/controllerregistration/controllerregistration_test.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		log = logger.MustNewZapLogger(logger.DebugLevel, logger.FormatJSON, logzap.WriteTo(GinkgoWriter))
+		virtualClient = fakeclient.NewClientBuilder().WithScheme(operatorclient.VirtualScheme).Build()
+		controllerRegistration = New(&record.FakeRecorder{})
+
+		extensionName = "test-extension"
+		ociRef = "local-extension:v1.2.3"
```
Comment (MartinWeindel) [inline pkg/operator/controller/extension/controllerregistration/controllerregistration_test.go]: Use name of extension
```suggestion
		ociRef = "test-extension:v1.2.3"
```

**Review by MartinWeindel — COMMENTED** [review]
lgtm, very well structured!

I only have some nit suggestions.

File: pkg/operator/controller/extension/reconciler.go (hunk 1 above)
Comment (timuthy) [reply pkg/operator/controller/extension/reconciler.go]: Thanks. I moved the log entry as well, to be more precise about deploying and deleting.

**Comment by oliver-goetz** [comment]
Looks very good, thanks 🚀 
/approve

lgtm for @MartinWeindel

**Pull request merged**